### PR TITLE
Support individual exemption deregistration

### PR DESCRIPTION
**Pull request by eminnett**
https://eaflood.atlassian.net/browse/RUBY-63

Add functionality to allow back office users to deregister individual exemptions. The feature includes the following functionality:

- users can choose to ‘cease’ or ‘revoke’ an exemption
- users must provide a reason before deregistering
- only active exemptions can be deregistered
- only super agents can do this

### CODE REVIEW

File: app/controllers/deregister_exemptions_controller.rb [hunk 1]
```diff
@@ -39,4 +34,11 @@ def handle_permission_denied
     unsuccessful_redirection = WasteExemptionsEngine::ApplicationController::UNSUCCESSFUL_REDIRECTION_CODE
     redirect_to "/pages/permission", status: unsuccessful_redirection
   end
+
+  def process_form_submission(deregistration_service)
+    deregistration_service.deregister!(@deregister_exemptions_form.state_transition)
+    @resource.update_attributes(deregistration_message: @deregister_exemptions_form.message)
```
Comment (irisfaraway) [inline app/controllers/deregister_exemptions_controller.rb]: I think traditionally we've done these updates in the form object following a successful `valid?` check, so I'd lean towards doing that again in this case for consistency's sake - is there a reason to do it in the controller instead?

**Comment by eminnett** [comment]
I've realised the old system includes an "exemption_deregister_date". I will add this column to the DB as well as set it via an AASM call back.

I will also add an AASM callback to set the `expired_on` column.

File: spec/models/registration_exemption_spec.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  describe "exemption state" do
+    context "when the state is 'active'" do
+      before(:each) { registration_exemption.state = :active }
+
+      it "can transition to any of the inactive states" do
+        expect(permitted_states).to match_array(inactive_states)
+      end
+
+      transitions.zip(inactive_states).each do |tansition, expected_state|
```
[the remark this answers is not in the document]
Comment (eminnett) [reply spec/models/registration_exemption_spec.rb]: Fixed in a4259f9.

File: config/locales/forms/deregister_exemptions.en.yml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+en:
+  deregister_exemptions:
+    new:
+      title: "Deregister Exemptions"
+      heading: "Deregister Exemption %{exemption_code} for Registration %{registration_reference}"
+      labels:
+        state_transition: "Deregistration Action"
+        state_transition_hint: "The action sets the state of the exemption regsitration."
```
[the remark this answers is not in the document]
Comment (eminnett) [reply config/locales/forms/deregister_exemptions.en.yml]: Fixed in a4259f9.

File: app/forms/deregister_exemptions_form.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+  attr_accessor :state_transition, :message
+
+  def submit(params)
+    self.state_transition = params[:state_transition]
+    self.message = params[:message]&.strip
+    valid?
+  end
+
+  def state_tranistion_options
```
[the remark this answers is not in the document]
Comment (eminnett) [reply app/forms/deregister_exemptions_form.rb]: Fixed in a4259f9.

File: spec/forms/deregister_exemptions_form_spec.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  describe "#submit" do
+    context "when the form is valid" do
+      it "should submit" do
+        expect(form.submit(state_transition: "revoke", message: "This exemption is no longer relevant")).to eq(true)
+      end
+    end
+
+    context "when the form is not valid" do
+      it "should not submit" do
+        expect(form.submit(state_transition: "deactivate", message: "X" * 501)).to eq(false)
```
[the remark this answers is not in the document]
Comment (eminnett) [reply spec/forms/deregister_exemptions_form_spec.rb]: Fixed in 91d495d.

File: app/controllers/deregister_exemptions_controller.rb (hunk 1 above)
Comment (eminnett) [reply app/controllers/deregister_exemptions_controller.rb]: Fixed in 1e73e82.

File: app/controllers/deregister_exemptions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    end
+  end
+
+  private
+
+  def find_resource(id)
+    @resource = WasteExemptionsEngine::RegistrationExemption.find(id)
+  end
+
+  def handle_permission_denied
```
[the remark this answers is not in the document]
Comment (eminnett) [reply app/controllers/deregister_exemptions_controller.rb]: Fixed in 8c673bf.

File: app/services/deregistration_service.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+class DeregistrationService
+  ALLOWED_ROLES = %w[system super_agent].freeze
```
[the remark this answers is not in the document]
Comment (eminnett) [reply app/services/deregistration_service.rb]: Fixed in 8c673bf.

**Comment by eminnett** [comment]
@irisfaraway I had to add some code to the `User` model  to be able to call `can?` on the current user inside the de-registration service (commit ed42b41). Are you happy with how this is done?

File: app/views/shared/_resource_exemptions_list.html.erb [hunk 8]
```diff
@@ -29,7 +29,7 @@
         <% end %>
       </td>
       <td class="govuk-table__cell">
-        <% if DeregistrationService.new(current_user, re).deregistration_allowed? %>
+        <% if can? :dergister, re %>
```
Comment (irisfaraway) [inline app/views/shared/_resource_exemptions_list.html.erb]: Typo spotted! `dergister`

File: app/models/concerns/waste_exemptions_engine/can_deactivate_exemption.rb [hunk 9]
```diff
@@ -15,19 +15,27 @@ module CanDeactivateExemption
...
         end
 
         event :expire do
           transitions from: :active,
                       to: :expired
         end
       end
+
+      # Transition effects
+      def deregister_exemption
```
Comment (irisfaraway) [inline app/models/concerns/waste_exemptions_engine/can_deactivate_exemption.rb]: I wonder if a name like `update_deregistered_on` would be a bit clearer? My initial assumption of what the method does based on this name would be that it was applying the state change.

**Comment by eminnett** [comment]
The AWS related commits are simply the product of me trying to debug the S3 errors in Travis. Since the errors only showed up in Travis, I had to push the trial and error fixes. None of them worked so reverted the changes.

Next time I'll do that debugging in a separate branch. Hopefully pull/105 will fix the problem as the errors aren't showing up there.

**Comment by irisfaraway** [comment]
@eminnett Oops, yes, can see they were getting reverted now (time for more coffee, I think). I'm OK with debugging happening on the same branch but I guess my preference is for removing the commits afterwards. Fingers crossed that 105 solves it though 🤞

**Comment by eminnett** [comment]
No problem at all.

File: app/views/shared/_resource_exemptions_list.html.erb (hunk 8 above)
Comment (eminnett) [reply app/views/shared/_resource_exemptions_list.html.erb]: Fixed in 21e29c8.

File: app/models/concerns/waste_exemptions_engine/can_deactivate_exemption.rb (hunk 9 above)
Comment (eminnett) [reply app/models/concerns/waste_exemptions_engine/can_deactivate_exemption.rb]: Fixed in 3c6532f.

**Pull request merged**